### PR TITLE
Add support for importing PKCS12 files.

### PR DESCRIPTION
**Pull request by fatmcgav**
Added ':source_password' param to provide PKCS12 password.
Added tests to support.

- [x] Clean up debug statements
- [x] Add Acceptance tests

### CODE REVIEW

**Comment by fatmcgav** [comment]
@hunner / @eputnam I wanted to raise this PR early to get some feedback on the changes, whilst I'm still working on testing etc...

All the current unit & acceptance tests pass, and the few unit tests I've added pass too... 
It also works in anger in local manifest testing... 

So yeh, comments welcome...

File: lib/puppet/provider/java_ks/keytool.rb [hunk 1]
```diff
@@ -54,19 +54,33 @@ def get_password
 
   def password_file
     pword = get_password
+    Puppet.debug("Pword = #{pword.inspect}")
```
Comment (hunner) [inline lib/puppet/provider/java_ks/keytool.rb]: I obviously wouldn't print the passwords in logs in the final version, as well as most of the other debug statements.

File: lib/puppet/type/java_ks.rb [hunk 2]
```diff
@@ -149,6 +149,10 @@ def value=(*values)
     defaultto 120
   end
 
+  newparam(:source_password) do
+    desc "The source keystore password"
```
Comment (hunner) [inline lib/puppet/type/java_ks.rb]: Is this only valid when `storetype => 'pkcs12'`? If so, it should do some validation as such.

File: lib/puppet/provider/java_ks/keytool.rb (hunk 1 above)
Comment (fatmcgav) [inline lib/puppet/provider/java_ks/keytool.rb]: Yeh, I'll tidy the debug lines up when I'm finished testing :)

File: lib/puppet/type/java_ks.rb (hunk 2 above)
Comment (fatmcgav) [inline lib/puppet/type/java_ks.rb]: Yep, only required for `pkcs12`... Will add some validation as such...

**Comment by tphoney** [comment]
HI @fatmcgav! looks like this now needs a rebase. @hunner would you be happy to merge, once the rebase is fixed ?

File: lib/puppet/type/java_ks.rb (hunk 2 above)
Comment (fatmcgav) [inline lib/puppet/type/java_ks.rb]: Validation added in 22e193 :)

**Comment by fatmcgav** [comment]
@tphoney I've rebased and also addressed a couple of the other comments that @hunner made. 

If both happy with the code, then I can get it squashed down into a single commit so that I can be merged in :)

**Comment by tphoney** [comment]
@fatmcgav It looks like the acceptance tests are failing. can you have a look. please reach out, if you need a hand.

**Comment by fatmcgav** [comment]
@tphoney So I've fixed one error, but the test is still failing :( 

Now getting: 
```
Error: Execution of 'keytool -importkeystore -srcstoretype PKCS12 -destkeystore /etc/pkcs12.ks -srckeystore /tmp/leaf.p12' returned 1: Enter destination keystore password:  Re-enter new password: Enter source keystore password:  keytool error: java.io.IOException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded
Error: /Stage[main]/Main/Java_ks[Leaf Cert:/etc/pkcs12.ks]/ensure: change from absent to latest failed: Execution of 'keytool -importkeystore -srcstoretype PKCS12 -destkeystore /etc/pkcs12.ks -srckeystore /tmp/leaf.p12' returned 1: Enter destination keystore password:  Re-enter new password: Enter source keystore password:  keytool error: java.io.IOException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded
```

However generating a cert in the same manner locally and then importing via Keytool works... 

So I'm slightly confused on that one... Google doesn't seem to turn up anything of use... :(

**Comment by fatmcgav** [comment]
Hah, found it... There was an incorrect password in spec test... 

Updated and the tests have now passed 🎉 

So, next steps... Squash it down?

**Comment by tphoney** [comment]
Excellent work as usual @fatmcgav  !!!

**Comment by fatmcgav** [comment]
Cheers @tphoney 

Any chance of a new release anytime soon? As keen to use https://github.com/puppetlabs/puppetlabs-java_ks/pull/151 aswell...

**Comment by tphoney** [comment]
I have created a ticket for release. Hopefully soon!

**Comment by edestecd** [comment]
This feature is not documented in the README

**Comment by ekohl** [comment]
It looks like this means you can no longer import PEM encoded files into a PKCS12 store. Was this ever officially supported?

**Pull request merged**